### PR TITLE
Combine all ``rec()`` related to ``Outer``

### PR DESCRIPTION
**Pull request by Li-Xiang-Ideal**
``rec()`` and ``rec_sparse()`` in ``Outer`` is combined into ``unpack_outer``. Now anyone can write a new tensor-product-like function with it.

I found it a tough task to name the variables/functions clearly while avoiding too many line breaks caused by isort & black. So any suggestions on variable/function naming is welcomed.

Also this PR is somewhat of a test since it seems I still don't have write access.

### CODE REVIEW

**Comment by Li-Xiang-Ideal** [comment]
I do not have write access. But if that's not an easy issue to resolve then it doesn't matter, you may just merge my PR after review.

File: mathics/eval/tensors.py [hunk 1]
```diff
@@ -76,6 +74,85 @@ def get_dimensions(expr, head=None):
...
+    )
+    ```
+    """
+    (
+        cond_next_list,  # return True when the next list should be unpacked
+        get_elements,  # get elements of list, tuple, ListExpression, etc.
+        apply_head,  # e.g. lambda elements: Expression(head, *elements)
+        apply_f,  # e.g. lambda current: Expression(f, *current)
+        join_elem,  # join current lowest level elements (i.e. current) with a new one
+        if_nested,  # Ture for result as nested list ({{a,b},{c,d}}), False for result as flattened list ({a,b,c,d}})
```
Comment (rocky) [inline mathics/eval/tensors.py:130]: Ture -> True

**Comment by rocky** [comment]
This is a great and much needed. I will look at this and the access issue when I have a chance.

File: mathics/eval/tensors.py (hunk 1 above)
Comment (Li-Xiang-Ideal) [reply mathics/eval/tensors.py]: Ah, yeah. Fixed it.

**Comment by Li-Xiang-Ideal** [comment]
> > I do not have write access. But if that's not an easy issue to resolve then it doesn't matter, you may just merge my PRs after review.
> 
> I think there were a couple more buttons I needed to press here. You should have received an invitation now. You will also need to set up two-factor authentication (2FA).

Received and joined.

**Comment by rocky** [comment]
> If you are referring to repository Mathics3/mathics-core instead of Li-Xiang-Ideal/mathics-core, than I found myself still unable to create a new branch in Mathics3/mathics-core.

Check your `.git/config` file and there should be line in it like this: 

```
[remote "origin"]
	url = git@github.com:Mathics3/mathics-core.git
	fetch = +refs/heads/*:refs/remotes/origin/*
```

I am guessing you tried to do this from a checked out repository from before you were a member, so the configuration has read-only access enabled.  If you see "https://github.com/Mathics3/mathics-core.git" you set things up read only.

Another way to check if you have read-write access is to try editing from the we UI.

**Comment by Li-Xiang-Ideal** [comment]
> The essential part, the URL is missing in the screenshot above.
> 
> When you go to this link: https://github.com/Mathics3/mathics-core/edit/master/README.rst do you still see that blue warning?

It appears anyway

image

> What does
> 
> ```
> git clone git@github.com:Mathics3/mathics-core.git
> ```
> 
> from a command line do?

```
$ git clone git@github.com:Mathics3/mathics-core.git
Cloning into 'mathics-core'...
remote: Enumerating objects: 54236, done.
remote: Counting objects: 100% (2525/2525), done.
remote: Compressing objects: 100% (946/946), done.
remote: Total 54236 (delta 1742), reused 2178 (delta 1572), pack-reused 51711
Receiving objects: 100% (54236/54236), 26.78 MiB | 1.74 MiB/s, done.
Resolving deltas: 100% (41087/41087), done.
```

**Comment by rocky** [comment]
> > The essential part, the URL is missing in the screenshot above.
> > When you go to this link: https://github.com/Mathics3/mathics-core/edit/master/README.rst do you still see that blue warning?
> 
> It appears anyway
> 
> image
> 
> > What does
> > ```
> > git clone git@github.com:Mathics3/mathics-core.git
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > from a command line do?
> 
> ```
> $ git clone git@github.com:Mathics3/mathics-core.git
> Cloning into 'mathics-core'...
> remote: Enumerating objects: 54236, done.
> remote: Counting objects: 100% (2525/2525), done.
> remote: Compressing objects: 100% (946/946), done.
> remote: Total 54236 (delta 1742), reused 2178 (delta 1572), pack-reused 51711
> Receiving objects: 100% (54236/54236), 26.78 MiB | 1.74 MiB/s, done.
> Resolving deltas: 100% (41087/41087), done.
> ```
> 

This suggests that you have write access to the repository in that cloned directory.  If you want to use the previously checked-out directory, you will need to adjust `.git/config` and the line under `[remote "origin"]` line as mentioned before.

**Comment by Li-Xiang-Ideal** [comment]
Close this PR to make it more convenient to try some new unit tests for ``unpack_outer()`` in ``Li-Xiang-Ideal/mathics-core/tree/master/test/eval/test_tensor.py``

**Pull request closed without merging**